src/lazy/encoder/binary/value_writer.rs

```diff
@@ -0,0 +1,886 @@
+        const LIST_TYPE_CODE: u8 = 0xB0;
+        BinaryListWriter_1_0::new(BinaryContainerWriter_1_0::new(
+            LIST_TYPE_CODE,
+            self.allocator,
+            self.encoding_buffer,
+        ))
+    }
+
+    fn sexp_writer(&mut self) -> IonResult<BinarySExpWriter_1_0<'_, 'top>> {
+        const SEXP_TYPE_CODE: u8 = 0xC0;
+        Ok(BinarySExpWriter_1_0::new(BinaryContainerWriter_1_0::new(
+            SEXP_TYPE_CODE,
+            self.allocator,
+            self.encoding_buffer,
+        )))
+    }
+
+    fn struct_writer(&mut self) -> IonResult<BinaryStructWriter_1_0<'_, 'top>> {
+        const STRUCT_TYPE_CODE: u8 = 0xD0;
+        Ok(BinaryStructWriter_1_0::new(BinaryContainerWriter_1_0::new(
+            STRUCT_TYPE_CODE,
+            &self.allocator,
+            self.encoding_buffer,
+        )))
+    }
+
+    fn write_list<
+        F: for<'a> FnOnce(&mut <Self as ValueWriter>::ListWriter<'a>) -> IonResult<()>,
+    >(
+        mut self,
+        list_fn: F,
+    ) -> IonResult<()> {
+        self.list_writer().write_values(list_fn)
+    }
+    fn write_sexp<
+        F: for<'a> FnOnce(&mut <Self as ValueWriter>::SExpWriter<'a>) -> IonResult<()>,
+    >(
+        mut self,
+        sexp_fn: F,
+    ) -> IonResult<()> {
+        self.sexp_writer()?.write_values(sexp_fn)
+    }
+    fn write_struct<
+        F: for<'a> FnOnce(&mut <Self as ValueWriter>::StructWriter<'a>) -> IonResult<()>,
+    >(
+        mut self,
+        struct_fn: F,
+    ) -> IonResult<()> {
+        self.struct_writer()?.write_fields(struct_fn)
+    }
+}
+
+impl<'value, 'top> Sealed for BinaryValueWriter_1_0<'value, 'top> {}
+
+impl<'value, 'top> ValueWriter for BinaryValueWriter_1_0<'value, 'top> {
+    type ListWriter<'a> = BinaryListValuesWriter_1_0<'a>;
+    type SExpWriter<'a> = BinarySExpValuesWriter_1_0<'a>;
+    type StructWriter<'a> = BinaryStructFieldsWriter_1_0<'a>;
+
+    delegate! {
+        to self {
+            fn write_null(self, ion_type: IonType) -> IonResult<()>;
+            fn write_bool(self, value: bool) -> IonResult<()>;
+            fn write_i64(self, value: i64) -> IonResult<()>;
+            fn write_int(self, value: &Int) -> IonResult<()>;
+            fn write_f32(self, value: f32) -> IonResult<()>;
+            fn write_f64(self, value: f64) -> IonResult<()>;
+            fn write_decimal(self, value: &Decimal) -> IonResult<()>;
+            fn write_timestamp(self, value: &Timestamp) -> IonResult<()>;
+            fn write_string<A: AsRef<str>>(self, value: A) -> IonResult<()>;
+            fn write_symbol<A: AsRawSymbolTokenRef>(self, value: A) -> IonResult<()>;
+            fn write_clob<A: AsRef<[u8]>>(self, value: A) -> IonResult<()>;
+            fn write_blob<A: AsRef<[u8]>>(self, value: A) -> IonResult<()>;
+            fn write_list<F: for<'a> FnOnce(&mut Self::ListWriter<'a>) -> IonResult<()>>(
+                self,
+                list_fn: F,
+            ) -> IonResult<()>;
```

Comment:
> but it's probably not possible because of the `Self::`

I think it's possible by taking `Self` as a type parameter to the alias. Unfortunately, [trait aliasing is not yet stable](https://doc.rust-lang.org/unstable-book/language-features/trait-alias.html).

That said, we could do it by defining a new trait if we felt strongly enough:

```rust
// Empty trait that extends -------vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
trait WriteListFn<V: ValueWriter>: for<'a> FnOnce(&mut V::ListWriter<'a>) -> Result<(), ()> {
    // No methods of its own
}

trait ValueWriter: Sized {
    type ListWriter<'a>;

    fn write_list(mut self, list_fn: impl WriteListFn<Self>) -> Result<(), ()> {
        todo!()
    }
}
```
[[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=18674fb13e1534dbfd77d0c566fb69e4)]

Reply:
That's neat. I'll leave it up to you to decide if this is actually worth doing.

Reply:
I'm going to punt on this for now.